src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs

```diff
@@ -111,53 +116,119 @@ public virtual bool ShouldFormatSnippet(SnippetInfo snippetInfo)
+                    _languageGuidForSnippets,
+                    0, // shortCutOnly
+                    Array.Empty<string>(), // types
+                    0, // countTypes
+                    1, // includeNULLTypes
+                    1 // includeDulicates: Allows snippets with the same title but different shortcuts
+                    ).ConfigureAwait(false);
+
+                // The rest of the process requires being on the UI thread, see the explanation on 
+                // PopulateSnippetCacheFromExpansionEnumeration for details
+                await Task.Factory.StartNew(() => PopulateSnippetCacheFromExpansionEnumeration(expansionEnumerator),
+                    CancellationToken.None,
+                    TaskCreationOptions.None,
+                    ForegroundTaskScheduler).ConfigureAwait(false);
+            }
+            catch (RuntimeBinderException)
+            {
+                // The IExpansionManager.EnumerateExpansionsAsync could not be found. Use 
+                // IVsExpansionManager.EnumerateExpansions instead, but from the UI thread.
+                await Task.Factory.StartNew(() => PopulateSnippetCacheOnForeground(),
+                            CancellationToken.None,
+                            TaskCreationOptions.None,
+                            ForegroundTaskScheduler).ConfigureAwait(false);
             }
+        }
 
-            return SpecializedCollections.EmptyList<SnippetInfo>();
+        /// <remarks>
+        /// Changes to the <see cref="IVsExpansionManager.EnumerateExpansions"/> invocation
+        /// should also be made to the IExpansionManager.EnumerateExpansionsAsync
+        /// invocation in <see cref="PopulateSnippetCacheOnBackgroundWithForegroundFallback"/>.
+        /// </remarks>
+        private void PopulateSnippetCacheOnForeground()
+        {
+            AssertIsForeground();
+
+            IVsExpansionEnumeration expansionEnumerator = null;
+            _expansionManager.EnumerateExpansions(
+                _languageGuidForSnippets,
+                fShortCutOnly: 0,
+                bstrTypes: null,
+                iCountTypes: 0,
+                fIncludeNULLType: 1,
+                fIncludeDuplicates: 1, // Allows snippets with the same title but different shortcuts
+                pEnum: out expansionEnumerator);
+
+            PopulateSnippetCacheFromExpansionEnumeration(expansionEnumerator);
         }
 
-        private bool TryGetVsSnippets(out IVsExpansionEnumeration expansionEnumerator)
+        /// <remarks>
+        /// This method must be called on the UI thread because it eventually calls into
+        /// IVsExpansionEnumeration.Next, which must be called on the UI thread due to an issue
+        /// with how the call is marshalled.
+        /// 
+        /// The second parameter for IVsExpansionEnumeration.Next is defined like this:
```

Comment:
Is all of the work already done when `EnumerateExpansionsAsync` is complete?  These `Next` calls will be fast?